main.cpp

```diff
@@ -7448,49 +7450,57 @@ int main(int argc, char **argv) {
+                            bool had_note = false;
+                            fos << missing_device_string(tries>0, selected_cmd->requires_rp2350());
+                            if (tries > 0) {
+                                fos << " It is possible the device is not responding, and will have to be manually entered into BOOTSEL mode.\n";
+                                had_note = true; // suppress "but:" in this case
                             }
-                        };
-#if defined(__linux__) || defined(__APPLE__)
-                        printer(dr_vidpid_bootrom_cant_connect,
-                                " appears to be a RP2040 device in BOOTSEL mode, but picotool was unable to connect. Maybe try 'sudo' or check your permissions.");
-#else
-                        printer(dr_vidpid_bootrom_cant_connect,
-                                " appears to be a RP2040 device in BOOTSEL mode, but picotool was unable to connect. You may need to install a driver via Zadig. See \"Getting started with Raspberry Pi Pico\" for more information");
-#endif
-                        printer(dr_vidpid_picoprobe,
-                                " appears to be a RP2040 PicoProbe device not in BOOTSEL mode.");
-                        printer(dr_vidpid_micropython,
-                                " appears to be a RP2040 MicroPython device not in BOOTSEL mode.");
-                        if (selected_cmd->force_requires_pre_reboot()) {
-#if defined(_WIN32)
-                            printer(dr_vidpid_stdio_usb,
-                                    " appears to be a RP2040 device with a USB serial connection, not in BOOTSEL mode. You can force reboot into BOOTSEL mode via 'picotool reboot -f -u' first.");
-#else
-                            printer(dr_vidpid_stdio_usb,
-                                    " appears to be a RP2040 device with a USB serial connection, so consider -f (or -F) to force reboot in order to run the command.");
-#endif
+                            fos << "\n";
+                            fos.first_column(0);
+                            fos.hanging_indent(4);
+                            auto printer = [&](enum picoboot_device_result r, const string &description) {
+                                if (!had_note && !devices[r].empty()) {
+                                    fos << "\nbut:\n\n";
+                                    had_note = true;
+                                }
+                                for (auto d : devices[r]) {
+                                    fos << bus_device_string(std::get<1>(d)) << description << "\n";
+                                }
+                            };
+    #if defined(__linux__) || defined(__APPLE__)
+                            printer(dr_vidpid_bootrom_cant_connect,
+                                    " appears to be a RP2040 device in BOOTSEL mode, but picotool was unable to connect. Maybe try 'sudo' or check your permissions.");
```

Comment:
Are these error messages definitely only applicable to RP2040, or do they apply to RP2350 too?

Reply:
They're for both, but that's outside the scope of this PR (I think there are a few other places where it'd need fixing too) - this was just an indentation change